numpyro/contrib/nn/auto_reg_nn.py

```diff
@@ -38,35 +35,29 @@ def create_mask(input_dim, hidden_dims, permutation, output_dim_multiplier):
     :type output_dim_multiplier: int
     """
     # Create mask indices for input, hidden layers, and final layer
-    var_index = onp.zeros(permutation.shape[0])
-    var_index[permutation] = onp.arange(input_dim)
-    dtype = var_index.dtype
+    var_index = np.zeros(permutation.shape[0])
+    var_index = ops.index_update(var_index, permutation, np.arange(input_dim))
 
     # Create the indices that are assigned to the neurons
     input_indices = 1 + var_index
-    hidden_indices = [sample_mask_indices(input_dim, h) for h in hidden_dims]
-    output_indices = onp.tile(var_index + 1, output_dim_multiplier)
+    hidden_indices = [sample_mask_indices(input_dim - 1, h) for h in hidden_dims]
```

Comment:
why `- 1`?

Reply:
In section 4 of [MADE](https://arxiv.org/pdf/1502.03509.pdf) paper, these numbers `m(k)` lie between `1` and `D - 1` (the long explanation there is a unit with `m(k) = D` will depend on all inputs, hence could not be used to model `p(x_k | x_{i < k})`

Reply:
ok but please help me understand what is the exact claim you're making. there are three kinds of mask code you can imagine:

i) one that is correct but has too many zeros
ii) one that is correct and is somehow optimal in that it has the fewest possible zeros
iii) one that is incorrect because it's not autoregressive (in which case the inversion tests etc should fail)

in making this change are you moving from i) to ii) or what?

is @stefanwebb's implementation of masking in pyro somehow suboptimal? or is it just that his code is organized differently from the description in the MADE paper (but is ultimately equivalent) and you're making a change that makes it look more like the MADE paper (but doesn't functionally change anything).

Reply:
@martinjankowiak I'll try to make it clearer but please let me know if there are any vague point to you. :)

The change is just for i) to ii). As long as the masks created by `output_indices = (1, ..., D) > hidden_indices >= input_indices = (1, ..., D)`, autoregressive property is guaranteed. In the previous version, hidden indices are created equally from `1 to D` (in `sample_mask_indices`), then -1 (indeed, `-1` is involved by changing the above inequality to `output_idx >= hidden_idx > input_idx`). Equivalent, the previous PR samples hidden indices from 0 -> D - 1. According the the paper, these hidden nodes with `index = 0` will be constant (because that index will always be smaller than input_indices, the hidden node will be masked out). I think that also explains for the phrase *The output first in the order (e.g. x_2 in the figure) is connected to hidden units rather than being unattached. Tracing a path back through the network, however, this variable will still be unconnected to any input variables* because any output indices will be able to connect to that `idx=0` hidden node (`[1, ..., D] > 0`), but that hidden node cannot connect to any input node (`0 < [1, ..., D]`).

This PR samples hidden indices from (1 to D - 1) instead. So that phenomenon (hidden idx = 0) won't happen.